task_manager/task_manager/task_client.py

```diff
@@ -185,17 +191,65 @@ def _request_canceling(self, timeout: float) -> List[GoalInfo]:
             )
             raise CancelTaskFailedError("Cancel request timed out.") from e
-        return future.result().goals_canceling
+        return future.result()
+
+    def _handle_cancel_response(self, response: CancelGoal.Response) -> None:
+        """Handles the response from the cancel request.
+
+        :raises CancelTaskFailedError: If the cancel request fails
+        """
+        # There seems to be a bug in rclpy, making the return code to be 0 (ERROR_NONE),
+        # no matter if the cancel was rejected or accepted. So checking instead if the
+        # goal is within the cancelling goals.
+        if response.return_code == CancelGoal.Response.ERROR_UNKNOWN_GOAL_ID:
+            self._node.get_logger().info(
+                f"Action server {self.task_specs.topic} did not recognize the goal id. "
+                f"Maybe server has restarted during the task execution and the goal no longer exists. "
+                f"Considering the task canceled."
+            )
+            if not self._result_future:
+                raise CancelTaskFailedError("Couldn't cancel the task result future since it does not exist!")
+            self._result_future.cancel()
+
+        elif response.return_code == CancelGoal.Response.ERROR_GOAL_TERMINATED:
+            self._node.get_logger().info(
+                f"Action server {self.task_specs.topic} did not accept to cancel the goal. "
+                f"Goal seems to have already finished. Considering the task canceled."
+            )
+            if not self._result_future:
+                raise CancelTaskFailedError("Couldn't cancel the task result future since it does not exist!")
+            self._result_future.cancel()
+
+        else:
+            if not self._goal_handle:
+                raise CancelTaskFailedError("Couldn't cancel the task, goal handle does not exist!")
+            goal_ids_cancelling = [goal_info.goal_id for goal_info in response.goals_canceling]
+            if self._goal_handle.goal_id not in goal_ids_cancelling:
+                self._node.get_logger().error(
+                    f"Couldn't cancel the task. Action server {self.task_specs.topic} did not "
+                    f"accept to cancel the goal."
+                )
+                raise CancelTaskFailedError("Couldn't cancel the task!")
 
     def _goal_done_cb(self, future: Future) -> None:
-        """Called when the Action Client's goal finishes. Updates the task status and notifies callbacks further the
-        other callbacks.
+        """Called when the Action Client's goal finishes. Updates the task status and invokes task done callbacks.
 
         :param future: Future object giving the result of the action call.
         """
+        if future.cancelled():
+            self.task_details.status = TaskStatus.CANCELED
+            self.task_details.result = self.task_specs.msg_interface.Result()
```

Comment:
Is this result setting to empty message necessary? Can action server send a result even though the action was cancelled? I guess this would then override that :thinking:

Reply:
In this case, we are not canceling the actual action (we cannot since the goal was not known by the server or had already finished), and hence also have not / cannot receive the result of the actual goal. Instead, we only canceled the `Future` object that was being waited on to receive the result. This results in the `cancelled()` flag being raised for the `Future` and its done-callbacks being executed, but its result is not set and thus will be `None` if we get here (unless we set it ourselves).

There is a very very small theoretical chance for a data race that could lead to us still having the result here after the cancel failed:

- we call cancel and the server reports `ERROR_GOAL_TERMINATED` because the goal has just finished,
- we receive the cancel result before the actual result of the goal and cancel the `Future`,
- we receive the result of the action (and set the `Future`'s result) while handling the result of the canceling, but before any of the `Future`'s done-callbacks are executed.

However, even in this case the result would not end up in the task details since this done callback is only ever executed once (the callbacks are stored separately and then cleared under the `Future`'s lock before they are invoked). The only way that the result could be set for the future after this is if we read the `cancelled()` flag before the execution of `Future.set_result()` reaches the setting of the cancelled status, and the result setting in `Future.set_result()` happens before this line where we could try to read it: a window of about 5-6 lines of code. So possible, but extremely unlikely and almost impossible to detect reliably, at least without accessing the `Future`'s protected lock and/or other hacks, which I personally don't see being worth the expected benefit.

Reply:
Ah, so this future is set to `cancelled` now only on our side in these special cases? So it won't be set during a normal cancellation, when the cancelling is actually successful? If that's the case, then it is all good. I agree also with the rest what you said.

Reply:
This is the result future, so its `set_result` is called when the action's result is received by the client. It does not matter whether the actions end result is aborted, succeeded or canceled, the future reaches the done state in all these cases.

Calling cancel on the `Future` object is not the same as canceling the action: when canceling the future, no signal whatsoever is sent to the server. What it simply does is that we stop waiting for the Future to be resolved (for the task that it is connected to to be done). The future's done callbacks get invoked to do any cleanup and end state handling specified, and the future then gets removed from the client's pending request list so that nothing will happen if the client suddenly does receive the actual result later (in our case this should almost never happen because we already received word from the server that the goal is not known or is already finished).

Reply:
Makes sense 👌